### ISSUE TITLE
request and response syntax

### ISSUE THREAD

**Issue by sugizo** [klen/muffin#64]
**objective**
can have api from cli
**e.g.**
**request.headers['api_key']**
`curl -X GET -i http://localhost:8000/api/http_api_key/instrument/ -H 'api_key: stifix'`
**request.query_params**
`curl -X PUT -i "http://localhost:8000/api/put/instrument/2?name=name0&strings=0"`
**request.form()**
`curl -X PUT -d "name=name1&strings=1" -i http://localhost:8000/api/put/instrument/2`
**request.json()**
`curl -X PUT -d '{"name": "name2", "strings": 2}' -i http://localhost:8000/api/put_json/instrument/2`

in starlette we can use request
**e.g.**

```
def test(request):
    #request
    request.headers['api_key']
    request.query_params
    request.form()
    request.json()

    #responses
    responses.PlainTextResponse(jwt_encode)
    responses.JSONResponse(rows)
```
what is equivalent of starlette request and response above in muffin?

thanks

**Comment by klen** [comment]
**Request** (more info in Muffin docs: https://klen.github.io/muffin/api.html#request)

```python
    # request.headers['api_key']
    request.headers['api_key']

    # request.query_params
    request.url.query

    # request.form()
    request.form()

    # request.json()
    request.json()

    # Note: the request method parses data automatically
    # so it returns json/formdata depends on request content type
    request.data()

```

**Response** (more info in Muffin docs: https://klen.github.io/muffin/api.html#responses)
```python
    # responses.PlainTextResponse(jwt_encode)
    muffin.ResponseText(jwt_encode)

    # responses.JSONResponse(rows)
    muffin.ResponseJSON(rows)

   # Note: muffin supports autodetect for response type
   # so you don't need to wrap your responses into `muffin.Response*` classes (btw it's optional)
   async def view(request):
       return {'it': 'will', 'be': 'returned', 'as': ['json', 'response']}
```

Don't hesitate to ask more questions if you have any.

**Comment by sugizo** [comment]
request.form() and request.data() not work
```
@app.route('/api/put/{tablename}/{rec_id}', methods = ['PUT'] )
async def put(request):
    tablename = request.path_params['tablename']
    rec_id = request.path_params['rec_id']
    if request.url.query:
        values = request.url.query
    elif await request.data():
        values = await request.data()
    #elif await request.form():
        #values = await request.form()
    row = db(db[tablename]._id == rec_id).validate_and_update(**values)
    db.commit()
    return row.as_dict()

"""
curl -X GET -i http://localhost:8000/api/no_cache/instrument/2
curl -X PUT -d "name=name1&strings=1" -i http://localhost:8000/api/put/instrument/2
curl -X GET -i http://localhost:8000/api/no_cache/instrument/2
"""
```
**return on server**
`INFO:     127.0.0.1:50343 - "PUT /api/put/instrument/2 HTTP/1.1" 500 Internal Server Error`
**error on curl**
```
HTTP/1.1 500 Internal Server Error
date: Thu, 01 Apr 2021 13:43:48 GMT
server: uvicorn
content-length: 28

Server got itself in trouble
```

**Comment by klen** [comment]
@sugizo For memory optimization, the Muffin doesn't keep ``request.body (form, json, data)`` after reading. 

You are trying to read ``request.data`` twice so it raises an error. To see the error in console start the application in ``debug`` mode (``app = Application(debug=True)``).

I rewrite your example:

```python
@app.route('/api/put/{tablename}/{rec_id}', methods=['PUT'] )
async def put(request):
    tablename = request.path_params['tablename']
    rec_id = request.path_params['rec_id']
    values = request.url.query or await request.data()   # You don't need to load request.form separately, request.data does the work for you
    row = db(db[tablename]._id == rec_id).validate_and_update(**values)
    db.commit()
    return row.as_dict()

```

Thank you for your feedback, I'll update the documentation to reflect these refinements

**Comment by sugizo** [comment]
test your rewritten code above
`curl -X PUT -d "name=name1&strings=1" -i http://localhost:8000/api/put/instrument/2`
**result**
success

`curl -X PUT -d '{"name": "name2", "strings": 2}' -i http://localhost:8000/api/put/instrument/2`
**result**
Server got itself in trouble

# conclusion
`await request.data()` not do the job to cover await request.form() and await request.json()

# objective
`await request.data()` do the job to cover await request.form() and await request.json()
so that no need to define it as two function
**e.g. work**
```
@app.route('/api/put/{tablename}/{rec_id}', methods = ['PUT'] )
async def put(request):
    tablename = request.path_params['tablename']
    rec_id = request.path_params['rec_id']
    values = request.url.query or await request.data()
    row = db(db[tablename]._id == rec_id).validate_and_update(**values)
    db.commit()
    return row.as_dict()

"""
curl -X GET -i http://localhost:8000/api/no_cache/instrument/2
curl -X PUT -i "http://localhost:8000/api/put/instrument/2?name=name0&strings=0"
curl -X GET -i http://localhost:8000/api/no_cache/instrument/2
curl -X PUT -d "name=name1&strings=1" -i http://localhost:8000/api/put/instrument/2
curl -X GET -i http://localhost:8000/api/no_cache/instrument/2
"""

@app.route('/api/put_json/{tablename}/{rec_id}', methods = ['PUT'] )
async def put_json(request):
    tablename = request.path_params['tablename']
    rec_id = request.path_params['rec_id']
    values = await request.json()
    row = db(db[tablename]._id == rec_id).validate_and_update(**values)
    db.commit()
    return row.as_dict()

"""
curl -X GET -i http://localhost:8000/api/no_cache/instrument/2
curl -X PUT -d '{"name": "name2", "strings": 2}' -i http://localhost:8000/api/put_json/instrument/2
curl -X GET -i http://localhost:8000/api/no_cache/instrument/2
"""

```

# question
how to achieve objective above using muffin way ?

thanks

**Comment by klen** [comment]
``request.data()`` checks a request's content-type header to define a request's type. For `application/json` it tries to load request body as JSON, for `application/x-www-form-urlencoded`, `multipart/form-data` it loads formdata otherwise text.

I've checked the command 

    curl -X PUT -d '{"name": "name2", "strings": 2}' -i http://localhost:8000/api/put/instrument/2

 and looks like it sends `"content-type":"application/x-www-form-urlencoded"` header. So the Muffin tries to implement request body as formdata.

The correct command to send JSON with curl ([stackoverflow](https://stackoverflow.com/questions/7172784/how-do-i-post-json-data-with-curl))

    curl -X PUT -d '{"name": "name2", "strings": 2}' --header "content-type: application/json" -i http://localhost:8000/api/put/instrument/2

> Server got itself in trouble

What kind of error does it show in debug mode for the case? I'm using `Muffin-0.64.8` and don't see any errors here.

**Comment by sugizo** [comment]
`curl -X PUT -d '{"name": "name2", "strings": 2}' -i http://localhost:8000/api/put/instrument/2`
**result**
```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/sugizo/miniconda3/envs/python3_test/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 396, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/Users/sugizo/miniconda3/envs/python3_test/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/Users/sugizo/miniconda3/envs/python3_test/lib/python3.8/site-packages/asgi_tools/app.py", line 150, in __call__
    await self.lifespan(request, receive, send)
  File "/Users/sugizo/miniconda3/envs/python3_test/lib/python3.8/site-packages/asgi_tools/middleware.py", line 36, in __call__
    return await self.app(scope, receive, send)
  File "/Users/sugizo/miniconda3/envs/python3_test/lib/python3.8/site-packages/asgi_tools/middleware.py", line 34, in __call__
    return await self.__process__(scope, receive, send)
  File "/Users/sugizo/miniconda3/envs/python3_test/lib/python3.8/site-packages/asgi_tools/app.py", line 73, in __process__
    response = await self.app(scope, receive, send)
  File "/Users/sugizo/miniconda3/envs/python3_test/lib/python3.8/site-packages/asgi_tools/app.py", line 167, in __process__
    response = await match.target(request)  # type: ignore
  File "/Users/sugizo/Downloads/muffin/music/exposed/api.py", line 173, in put
    row = db(db[tablename]._id == rec_id).validate_and_update(**values)
  File "/Users/sugizo/miniconda3/envs/python3_test/lib/python3.8/site-packages/pydal/objects.py", line 2738, in validate_and_update
    raise ValueError("No fields to update")
ValueError: No fields to update
INFO:     127.0.0.1:52092 - "PUT /api/put/instrument/2 HTTP/1.1" 500 Internal Server Error
```

`curl -X PUT -d '{"name": "name2", "strings": 2}' -H "content-type: application/json" -i http://localhost:8000/api/put/instrument/2`
**result**
work

yes, you right, thanks for detail explanation

one more thing is about decorator
**e.g.
the code**
`@app.route('/api/no_cache/{tablename}/', methods = ['GET'] )`

**execute in terminal
either**
`curl -X GET -i http://localhost:8000/api/no_cache/instrument`
**or**
`curl -X GET -i http://localhost:8000/api/no_cache/instrument/`

**result**
Nothing matches the given URI

in muffin to make the curl work on case above, must remove the last slash (before tablename variable) in the decorator
while in another python web framework doesnt need, they automatically trim the last slash on the decorator, e.g. starlette, sanic, etc

**Comment by klen** [comment]
> ValueError: No fields to update

Oh, looks like the error raises in your code when `values` is empty. Maybe you should check and return `ResponseError.BAD_REQUEST()` in this situation.

> one more thing is about decorator

Thank you for the report. There was a bug when [trim_last_slash](https://klen.github.io/muffin/configuration.html#default-application-options) configuration option is enabled. The bug has been fixed now, please update to `Muffin==0.64.10`
